backend/src/prereq_loader.ts

```diff
@@ -0,0 +1,250 @@
+let totalNumRequests = 0;
+
+/**
+ * Asynchronously adds prereqs to a Schedule.
+ * Does not do mutation.
+ * @param schedules the schedule to add prereqs to
+ */
+export async function addPrereqsToSchedules(
+  schedules: Schedule[]
+): Promise<Schedule[]> {
+  // set the total to zero
+  totalNumRequests = 0;
+
+  // the loader to use for building a
+  const loader = new DataLoader<SimpleCourse, PrereqQueryResult>(
+    (keys: SimpleCourse[]) => queryCoursePrereqData(keys)
+  );
+
+  // return the results
+  let results = await Promise.all(
+    schedules.map((sched: Schedule) => prereqifySchedule(sched, loader))
+  );
+
+  // log the total number of requests.
+  // console.log("total: " + totalNumRequests);
+  totalNumRequests = 0;
+
+  return results;
+}
+
+/**
+ * Loads prereqs for a given schedule
+ * @param schedule the schedule
+ * @param loader the loader
+ */
+async function prereqifySchedule(
+  schedule: Schedule,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<Schedule> {
+  // does not do mutation!
+  const newYearMap: { [key: number]: ScheduleYear } = {};
+
+  // convert each of the years
+  for (const year of schedule.years) {
+    newYearMap[year] = await prereqifyScheduleYear(
+      schedule.yearMap[year],
+      loader
+    );
+  }
+
+  // reconstructs a year
+  return {
+    years: schedule.years.slice(0),
+    yearMap: newYearMap,
+    id: schedule.id,
+  };
+}
+
+/**
+ * Asynchronously adds prereqs to a provided ScheduleYear
+ * @param yearObj the year object to add prereqs to
+ * @param loader the loader
+ */
+async function prereqifyScheduleYear(
+  yearObj: ScheduleYear,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleYear> {
+  return {
+    year: yearObj.year,
+    fall: await prereqifyScheduleTerm(yearObj.fall, loader),
+    spring: await prereqifyScheduleTerm(yearObj.spring, loader),
+    summer1: await prereqifyScheduleTerm(yearObj.summer1, loader),
+    summer2: await prereqifyScheduleTerm(yearObj.summer2, loader),
+    isSummerFull: yearObj.isSummerFull,
+  };
+}
+
+/**
+ * Asynchronously adds prereqs to a term's courses.
+ * @param termObj the term object to add prereqs to
+ * @param loader the loader
+ */
+async function prereqifyScheduleTerm(
+  termObj: ScheduleTerm,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleTerm> {
+  // the new classes.
+  const newClasses: Promise<ScheduleCourse[]> = Promise.all(
+    termObj.classes.map((course: ScheduleCourse) =>
+      prereqifyScheduleCourse(course, termObj.termId, loader)
+    )
+  );
+
+  return {
+    season: termObj.season,
+    year: termObj.year,
+    termId: termObj.termId,
+    id: termObj.id,
+    classes: await newClasses,
+    status: termObj.status,
+  };
+}
+
+/**
+ * Asynchronously adds prereqs to a provided course.
+ * @param course the course to add prereqs to
+ * @param loader the loader
+ */
+async function prereqifyScheduleCourse(
+  course: ScheduleCourse,
```

Comment:
I don't know if you already tried this but maybe you could `console.log(currentTime)` in here. This would tell us what the time intervals are between the calls to `prereqifyScheduleCourse` for each `ScheduleCourse` in our array. It could give us some insights on the number of queries in a single `DataLoader` batch.

Reply:
Seems to be a couple of milliseconds. 

The "want a course" means `prereqifyScheduleCourse` was called, while "request a course" means the load function `queryCoursePrereqData` was called.
![Screen Shot 2019-11-11 at 3 40 09 PM](https://user-images.githubusercontent.com/22948432/68619689-f4b17780-0499-11ea-8e57-d8065bf0e9c9.png)
![Screen Shot 2019-11-11 at 3 40 15 PM](https://user-images.githubusercontent.com/22948432/68619696-f713d180-0499-11ea-9dae-193ea4018079.png)
![Screen Shot 2019-11-11 at 3 40 23 PM](https://user-images.githubusercontent.com/22948432/68619697-f8dd9500-0499-11ea-95f4-2f52abc7d62c.png)